py/mpprint.c

```diff
@@ -58,62 +57,60 @@ int mp_print_str(const mp_print_t *print, const char *str) {
     return len;
 }
 
+// Efficiently print `count` chars looping over `str[len]`.
+static void print_strn_cycle(const mp_print_t *print, const char *str, size_t len, size_t count) {
+    while (count > len) {
+        print->print_strn(print->data, str, len);
+        count -= len;
+    }
+    if (count > 0) {
+        print->print_strn(print->data, str, count);
+    }
+}
+
+// lcm(4,5) to match stride for both comma and underscore grouping
+#define PAD_BUF_SIZE 20
+
 int mp_print_strn(const mp_print_t *print, const char *str, size_t len, unsigned int flags, char fill, int width) {
     int left_pad = 0;
     int right_pad = 0;
     int pad = width - len;
-    int pad_size;
     int total_chars_printed = 0;
-    const char *pad_chars;
-
-    if (!fill || fill == ' ') {
-        pad_chars = pad_spaces;
-        pad_size = sizeof(pad_spaces) - 1;
-    } else if (fill == '0') {
-        pad_chars = pad_zeroes;
-        pad_size = sizeof(pad_zeroes) - 1;
-    } else {
-        // Other pad characters are fairly unusual, so we'll take the hit
-        // and output them 1 at a time.
-        pad_chars = &fill;
-        pad_size = 1;
-    }
 
-    if (flags & PF_FLAG_CENTER_ADJUST) {
-        left_pad = pad / 2;
-        right_pad = pad - left_pad;
-    } else if (flags & PF_FLAG_LEFT_ADJUST) {
-        right_pad = pad;
-    } else {
-        left_pad = pad;
-    }
+    // alignment ensures fast memset
+    #if defined(_MSC_VER)
+    __declspec(align(4)) char pad_buf[PAD_BUF_SIZE];
+    #else
+    alignas(4) char pad_buf[PAD_BUF_SIZE];
```

Comment:
Maybe I'm mistaken but wasn't `alignas` introduced in C11?  I've had to revert changes in `py/` that weren't C99-clean before :)

Reply:
Turning that into a macro sounds like a sensible choice, `py/misc.h` may also be another candidate for its location.

But does `alignas` really speed things up?  It is my understanding that stack arrays should already be aligned to a machine word boundary unless some weird compiler flag is in effect that makes it pack stack variables.  Or at least it was on pretty much any recent platform I've worked with, please don't quote me on that.

It looks like `pad_buf` is operated on a byte-by-byte basis anyway except for the call to `memset`, which the compiler handles as an intrinsic call.  For 20 bytes worth of zeroes it will probably turn that function call into a pretty tight opcode sequence anyway.

Reply:
Ok, I can confirm there's no longer any differences between having it explicitly aligned vs not.